walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,145 @@
+    
+    // sets up timeSlots array; anywhere left true has all the requestAttendees available
+    boolean[] mandatoryTimeSlots = convertToTimeSlots(requestAttendees, events);
+    boolean[] allTimeSlots = convertToTimeSlots(allAttendees, events);    
+
+
+    ArrayList<Integer> possibleSlots = convertToSlots(mandatoryTimeSlots, numRequestSlots);
+    ArrayList<Integer> allSlots = convertToSlots(allTimeSlots, numRequestSlots);
+    
+    ArrayList<TimeRange> possibleTimes = convertToTimes(possibleSlots, numRequestSlots, numEvents);
+    ArrayList<TimeRange> allTimes = convertToTimes(allSlots, numRequestSlots, numEvents);
+
+    if (allTimes.isEmpty() && !possibleTimes.contains(TimeRange.WHOLE_DAY)) {
+        return possibleTimes;
+    }
+    return allTimes;
+  }
+  
+  public static boolean[] convertToTimeSlots (Collection<String> requestAttendees, Collection<Event> events) {
+    boolean[] timeSlots = new boolean[TOTAL_NUM_SLOTS];
+    Arrays.fill(timeSlots, Boolean.TRUE);
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      TimeRange when = event.getWhen();
+      int start = (when.start()) / 30;
+      int eventSlots = when.duration() / 30;
+
+      Iterator<String> eventAi = eventAttendees.iterator(); 
+      while(eventAi.hasNext()) {
+        String eventAttendee = eventAi.next();
+
+        Iterator<String> requestAi = requestAttendees.iterator(); 
+        while (requestAi.hasNext()) {
+          String requestAttendee = requestAi.next();
+          if (eventAttendee.equals(requestAttendee)) {
+            for (int i = 0; i < eventSlots; i++) {
+              timeSlots[start + i] = false;
+            }
+            break;
+          }
+        }
+      }
+    }
+    return timeSlots;
+  }
+
+  // converts boolean availability to list of available slots
+  public static ArrayList<Integer> convertToSlots (boolean timeSlots[], int numRequestSlots) {
+    ArrayList<Integer> slots = new ArrayList<Integer>();
+    boolean add;
+    for (int i = 0; i < TOTAL_NUM_SLOTS; i++) {
+      add = false;
+      if (timeSlots[i]) {
+        add = true;
+        // checks if the duration is longer than a single time slot
+        for (int j = 1; j < numRequestSlots; j++) {
+          if ((i+j) < timeSlots.length && !timeSlots[i+j]) {
+            add = false;
+          }
+        }
+      }
+      if (add) {
+        for (int j = 0; j < numRequestSlots; j++) {
+          slots.add(i+j);
+        }
+      }
+    }
+    return slots;
+  }
+
+  // converts list of available slots to a list of TimeRanges 
+  public static ArrayList<TimeRange> convertToTimes (ArrayList<Integer> slots, int numRequestSlots, int numEvents) {
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    // if slots size is zero
+    if (slots.size() == 0) {
+      return possibleTimes;
+    }
+
+    int start = slots.get(0)*30;
+
+    // if there's just enough room
+    if (numRequestSlots == slots.size()) {
```

Comment:
This is assuming that the preprocessing done in `convertToSlots` happens before calling this method. Maybe add a comment explaining your assumptions of the parameters at the top of the method? Something like "Assumes `slots` contains only time slots that are at least the length of the requested event?

Reply:
done :)